### ISSUE TITLE
General refactoring of reducers and actions

### ISSUE THREAD

**Issue by charbugs** [ProjectMirador/mirador#1926]
I like to make a general refacotring of our action and reducer implementation. 

### Context

As you will have noticed, things in the action and reducer files are getting more complex and we're starting to mixing up different patterns to accomplish things. So far, we mainly did state updates that only changes a certain state slice, e.g. changing canvas index of a window, adding a new manifest. But as, for example, PR #1922 shows there will be cases where we have to deal with actions that do more complex things, like closing a window, what implies removing the window and removing all the companion windows that belong to that window. To enable convenient implementations of such features we should apply more predictable patterns for state management.

### Problems with our reducers/actions

* For my understanding there are two general aproaches to handle state in redux. You can either put many logic in the reducer and keep action creators very simple, or the other way around, having only simple reducers and let the actions do the logic (the latter I prefer). We are currently not decided about which way to go and do sometimes so, sometimes so.

* There are different kind of ways to passing properties to action creators, sometimes single objects are required that hold the properties to process, sometimes the properties have to be passed as single parameters.

* In the reducers, the expressions to merge state without mutating it are sometimes very large and hard to read. With libraries like immutable.js this can be done with one-liners.

* The execution path of the application is hard to follow if we handle a certain action type in different reducers. I know that's a nifty feature of redux, but I think it's not a good one.

### Suggestion

I created a [proof of concept](https://github.com/charbugs/store-refactoring-fiddle) for a pattern that seems reasonable to me. The general idea is to make reducers very simple functions that can only add, update and remove the data they are responsible for. Also, for each reducer there is a set of basic action creators (add, update, remove) with an identical signature. On top of that there are combined/derived action creators that use this basic actions to accomblish more specific things to the UI. That is, there is a general predictable abstration of reducer logic. Actually, there should be no need to touch a reducer at all.

I'm currently figuring out if all use cases that we have in our code base so far can be accomplished with that pattern.

I would be happy about your reviews of the POC and comments on that.

**Comment by mejackreed** [comment]
I'd be interested to see where this goes. A bit of historical underpinnings to the diverse ActionTypes. We were originally thinking that adopters who had customized Mirador instances would subscribe and use the provided ActionTypes similar to how current Mirador adopters use the event system. While I like the simplicity of the approach in the proof of concept, do you think there would also be a way to provide additional ActionTypes for external usage?

For example, a Mirador adopter may have a customized version (or embedded application) that does something special when a sidebar is opened or closed. Using these ActionTypes was the original intent to provide this functionality. @aeschylus may have some additional context. With the create/update/delete approach, subscribers to actions would then need to perhaps dig around in the state change for the thing that they were looking for? Is there a better way to handle this?

I realize this use case is not fully defined yet, but wanted to share some of how we were thinking this could be used.

**Comment by charbugs** [comment]
Ok, I think I understand the problem. You mean with this pattern the redux store only ever receives this primitive add, update and remove actions and a plugin could not listen to a more specific action bescause they are encapsulated in the thunks?

On possible solution could be to fire special event actions in the thunks, like so:

```javascript
export function closeWindow (windowId) {
  return function (dispatch, getState) {
    const { companionIds } = getState().windows[windowId]
    companionIds.forEach(id => dispatch(removeCompanion(id)))
    dispatch(removeWindow(windowId))
    dispatch({ type: EventTypes.EVENT_WINDOW_CLOSED })
  }
}
```

Then the implementers could listen to this EventTypes, while the ActionTypes are internal to Mirador. This way you could precisely determine what events you want the outer world know about.

**Comment by charbugs** [comment]
I did some further improvements on the [POC](https://github.com/charbugs/store-refactoring-fiddle). 

As the reducer logic in this pattern is actually only boilerplate it's possible to create reducers automatically. That is, there is no reducer work to do at all.

Applied to our code base this means, that we could remove all reducer files and the corresponding tests.

**Comment by mejackreed** [comment]
Gotcha.. so then we are proposing moving all logic to the action creators. In theory this seems fine to me. But I'd love to hear others thoughts on this approach. cc @aeschylus @jeffreycwitt  I'm also curious how some of the "issues" with this will pan out: https://redux.js.org/recipes/using-immutablejs-with-redux#what-are-the-issues-with-using-immutablejs

I'm curious and will look into reading more about this. Thanks for your hard work on this.

**Comment by charbugs** [comment]
I'm not very attached to using immutable.js. It's just a convenient way to return state from the reducers. Hand crafted utility functions could do the the job too. 

I read the article about immutable.js and redux and I like to dispel the concerns about that: 

> Once you encapsulate your data with Immutable.JS, you have to use Immutable.JS’s get() or getIn() property accessors to access it.
> 
> This has the effect of spreading Immutable.JS across your entire codebase, including potentially your components, where you may prefer not to have such external dependencies. Your entire codebase must know what is, and what is not, an Immutable.JS object. It also makes removing Immutable.JS from your app difficult in the future, should you ever need to.
> 

The only place where immutable.js is used in the POC is in the reducers. Take for example the line `setIn(state, [action.id], action.payload)`. It's basically a shorthand for `fromJS(state).setIn([action.id], action.payload).toJson()`. That is, state is wrapped in a immutable.js object, processed and finally unwrapped to a plain javascript object. So we will have no immutable.js objects in state. 

As I said, hand crafted helper function could do the same job.

**Comment by charbugs** [comment]
I updated the [POC](https://github.com/charbugs/store-refactoring-fiddle) to automatically create basic actions that match the reducer needs.

**Comment by charbugs** [comment]
I just realized that this pattern is [described](https://redux.js.org/recipes/structuring-reducers/reusing-reducer-logic) in the Redux docs. They call it *higher-order reducers*. Also, there are a some [libraries](https://github.com/markerikson/redux-ecosystem-links/blob/master/action-reducer-generators.md) that address the reducer boilerplate problem. So, I think the risk of implementing this pattern to Mirador is low.

**Comment by mejackreed** [comment]
I'd be good with this approach, especially if we can tackle it piece by piece to hopefully share understanding with more of the team on it. One thing @sdellis and I were discussing yesterday was where the logic should take place for async or complex things. It seems like with your POC work "combinedActions" is a place where that type of  functionality happens. Perhaps we can still break out "combinedActions" into state organization pieces at first?

**Comment by charbugs** [comment]
Yes, I think it's possible and a good idea to apply this pattern incrementally, starting with the simple actions and move forward. 

I use the file name "combinedActions" to make the concept clear. When applied to Mirador, it should get split into several files.

**Comment by charbugs** [comment]
See #2035, #2049 for further discussion and [this wiki article](https://github.com/ProjectMirador/mirador/wiki/M3-State-Management:-Reducers,-Actions,-Events) for description of the pattern

**Comment by cbeer** [comment]
I think #3063 addressed the bulk of this?
